src/bin/tectonic/v2cli.rs

```diff
@@ -528,39 +534,66 @@ impl WatchCommand {
             .expect("Executable path wasn't valid UTF-8");
         let mut cmds = Vec::new();
         for x in self.execute.iter() {
-            let mut cmd = format!("{exe_name} -X ");
             let x = x.trim();
             if !x.is_empty() {
-                cmd.push_str(x);
+                let cmd = Command::Exec {
+                    prog: exe_name.clone(),
+                    args: vec!["-X".to_string(), x.to_string()],
+                };
                 cmds.push(cmd)
             }
         }
 
         if cmds.is_empty() {
-            cmds.push(format!("{exe_name} -X build"))
+            cmds.push(Command::Exec {
+                prog: exe_name,
+                args: vec!["-X".to_string(), "build".to_string()],
+            });
         }
 
-        let command = cmds.join(" && ");
-
-        let mut args = watchexec::config::ConfigBuilder::default();
-        let mut final_command = command.clone();
-        #[cfg(unix)]
-        final_command.push_str("; echo [Finished running. Exit status: $?]");
         #[cfg(windows)]
-        {
-            final_command.push_str(" & echo [Finished running. Exit status: %ERRORLEVEL%]");
-            args.shell(watchexec::Shell::Cmd);
-        }
-        #[cfg(not(any(unix, windows)))]
-        final_command.push_str(" ; echo [Finished running]");
+        let (shell, command) = (
+            Shell::Cmd,
+            "echo [Finished running. Exit status: %ERRORLEVEL%]",
+        );
+        #[cfg(unix)]
+        let (shell, command) = (
+            Shell::Unix("bash".to_string()),
+            "echo [Finished running. Exit status: $?]",
+        );
 
-        args.cmd(vec![final_command])
-            .paths(vec![env::current_dir()?])
-            .ignores(vec!["build".to_owned()])
-            .on_busy_update(OnBusyUpdate::Queue);
-        let args = args.build().map_err(NewError::from)?;
+        cmds.push(Command::Shell {
+            shell,
+            args: vec![],
+            command: command.to_string(),
+        });
+
+        let mut runtime_config = watchexec::config::RuntimeConfig::default();
+        runtime_config.commands(cmds);
+
+        let current_dir = env::current_dir()?;
+
+        let rt = runtime::Builder::new_current_thread().build()?;
+        let filter = rt
+            .block_on(GlobsetFilterer::new(
+                &current_dir,
+                [],
+                [("build".to_string(), None)],
+                [],
+                [],
+            ))
+            .unwrap();
+
+        runtime_config
+            .pathset([&current_dir])
+            .filterer(Arc::new(filter))
+            .on_pre_spawn(|pre_spawn: PreSpawn| async move {
+                println!("[Running `{}`]", pre_spawn.command);
+                Ok::<_, Infallible>(())
+            });
+
+        let exec_handler = Watchexec::new(InitConfig::default(), runtime_config);
```

Comment:
Whoops, alright, I'll fix it. Also, is there any infrastructure for adding an integration test for running a `watch` command? I'd love to add some.

Reply:
It looks like `executable.rs` is the file for such a test, I'm going to try to add one

Reply:
Before a full fix, looking at the error - this may be a bit of an issue, since fundamentally, `Watchexec` wants you to be doing it in an async environment. The actual thing spawned is a tokio task, so to properly use it, we need to either immediately wait on it, or make tectonic watch command async as a whole.

Reply:
Thinking about it more - it should be fine to block on it, right? The watch command doesn't exit till killed as far as I can tell, so it should be okay to just block and continue only when the task dies or we're stopped